### PR TITLE
feat(mcp): map JSON-Schema enum/const → Literal and anyOf/oneOf → Union for tool params

### PR DESCRIPTION
**Pull request by alexagr**
## Summary

`FastMCPClient._schema_to_field` turns each MCP tool parameter's JSON Schema into a pydantic field type. Two common constructs were being dropped, so the generated `ToolMessage` lost type information both for validation and in the schema the LLM sees:

- **`enum` / `const`** fell through to the base scalar type (`{"type": "string", "enum": [...]}` → plain `str`).
- **`anyOf` / `oneOf` / `allOf`** collapsed to `Any` with a warning — including the very common `anyOf: [T, {"type": "null"}]` that fastmcp emits for `Optional[...]` params.

This maps them to precise Python types.

## Changes (all in `_schema_to_field`)

**`enum` / `const` → `typing.Literal`** (before the `type` branches):

```python
enum_values = schema.get("enum", [schema["const"]] if "const" in schema else None)
if enum_values and all(isinstance(v, (str, int, bool)) or v is None for v in enum_values):
    literal_type = Literal[tuple(enum_values)]
    if not is_required:
        literal_type = Optional[literal_type]
    return literal_type, Field(default=default, description=desc)
```
Guarded to `Literal`-compatible scalars (`str`/`int`/`bool`/`None`); enums with floats or objects fall through to existing handling.

**`anyOf` / `oneOf` → `typing.Union`** (replacing the old `Any` fallback), converting each branch recursively; a `{"type": "null"}` branch — or an optional field — yields `Optional`:

```python
sub_schemas = schema.get("anyOf") or schema.get("oneOf")
if isinstance(sub_schemas, list) and sub_schemas:
    non_null = [s for s in sub_schemas if not (isinstance(s, dict) and s.get("type") == "null")]
    has_null = len(non_null) != len(sub_schemas)
    if non_null:
        member_types = tuple(self._schema_to_field(name, s, prefix, is_required=True)[0] for s in non_null)
        union_type = Union[member_types]
        if has_null or not is_required:
            union_type = Optional[union_type]
        return union_type, Field(default=default, description=desc)
```

**`allOf`**: a single subschema is converted directly; multi-schema intersections keep the `Any` fallback (no clean typing analogue). `oneOf` maps to `Union` too — typing has no exclusive-or.

## Why it matters

- **Validation**: out-of-set enum values and wrong-type union values now raise `ValidationError` instead of being silently accepted.
- **Better LLM grounding**: pydantic re-emits `enum`, `anyOf`, and nullability in `model_json_schema()`, so the model receives exact parameter types instead of "string"/"any".

## Testing

- `test_enum_field_becomes_literal`: a `Literal["celsius","fahrenheit"]` param (emitted as a schema `enum`) → `Literal` field; validates `"celsius"`, rejects `"kelvin"`, re-emits the `enum` in `model_json_schema()`.
- `test_anyof_maps_to_union_and_optional`: `int | str` → `Union[int, str]` (both validate); `Optional[int]` → `Union[int, None]` (accepts `None` and `int`).

### CODE REVIEW

File: langroid/agent/tools/mcp/fastmcp_client.py [hunk 1]
```diff
@@ -263,6 +275,22 @@ def _schema_to_field(
...
+        # (str/int/bool/None). This takes precedence over the plain `type`
+        # branches so the allowed values are preserved for validation and echoed
+        # back into the model's JSON schema (which the LLM sees). JSON-Schema
+        # enums may also hold floats or objects, which Literal rejects; those
+        # fall through to the type-based handling below.
+        enum_values = schema.get(
+            "enum", [schema["const"]] if "const" in schema else None
+        )
+        if enum_values and all(
+            isinstance(v, (str, int, bool)) or v is None for v in enum_values
```
[the remark this answers is not in the document]
Comment (alexagr) [reply langroid/agent/tools/mcp/fastmcp_client.py]: fixed

**Comment by alexagr** [comment]
@pchalasani this complete work in this branch - will appreciate if you could review the changes and merge them

**Comment by pchalasani** [comment]
Thanks so much for this, @alexagr — really solid work, and I appreciate you iterating on the codex feedback (the malformed-enum guard, the $ref resolution, and the Pydantic v2 `$defs` cleanup fix in particular).

I've taken it over in #1059, which builds directly on your commits (squashed into the first commit, authorship preserved). On top of your work I ran a full adversarial codex review loop over the whole branch — including the $ref-resolution and $defs commits that the GitHub codex bot never got to — and added a bit of hardening it surfaced:

- $ref resolution is now restricted to local `#/$defs/<Name>` pointers (a non-local or non-string `$ref` degrades to `Any` instead of mis-resolving on the trailing path segment);
- an `anyOf`/`oneOf` whose members are all `{"type": "null"}` now maps to `NoneType` rather than `Any`;
- extra tests for oneOf/const/allOf(single+multi)/integer-enum/non-local-$ref/null-only-union, and a short docs section on the parameter type mapping.

Closing this in favor of #1059. Thanks again for the contribution! 🙏
